### ISSUE TITLE
Credentail loading: Severe efficiency degredation since 3.0

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#18814]
This is a follow-up on #14837, which got closed due to a partial solution regarding the heavy use of heap memory allocation.

I've just been using the tool provided by @richsalz in #16540, compiled like this:
```
cd test/
gcc -g -I../include timing.c ../libcrypto.a -ldl -lpthread -o timing
```
When run on current 3.x master, I typically get figures like these:
```
./timing -wc timing-cert.pem
user      0 sec 57846 microsec
sys       0 sec 69877 microsec
```
and
```
./timing -wp timing-key.pem 
user      0 sec 95863 microsec
sys       0 sec 115708 microsec
```

Whereas with the latest 1.1.1 branch, I typically get for instance:
```
./timing -wc timing-cert.pem
user      0 sec 17389 microsec
sys       0 sec 17389 microsec
```
and
```
./timing -wp timing-key.pem
user      0 sec 14719 microsec
sys       0 sec 10223 microsec
```

BTW, on my Debian system times reported vary a lot between consecutive runs - I've copied some 'average' outcomes.

So with 1.1.1, cert and key loading appear reasonably efficient,
while since 3.0 loading a simple cert (including a public key) degraded by a factor around 3.5, 
and loading a private key degraded by a factor of more than 6.

@t8m wrote in https://github.com/openssl/openssl/issues/14837#issuecomment-862506046:
> The obvious reason being the change from legacy to provider based parsing of the public keys in certificates.

Apparently the inefficiencies are due to generalizing the loading mechanism using the provider mechanism and maybe the STORE API,
which appear pretty involved when having a look at the code and stepping through them while debugging.

Is this considered bearable?
Hopefully some bottlenecks can be identified and removed.

**Comment by beldmit** [comment]
See also #16791 and #17627

**Comment by paulidale** [comment]
It's mostly due to the generalisation of the loading -- checking many formats takes time.

**Comment by richsalz** [comment]
Who needed a general mechanism, anyway?  And who decided the generalization was a good thing to put into the common path?  Perhaps an internal API that allows functions to specify a search path would help reduce the pain of this generalization.

**Comment by DDvO** [comment]
Before 3.0 it was already possible to load keys of all known types in reasonable time.
Even if the extra flexibility of handling further (provider-defined) key types is the main reason for the credential loading taking 3-6 times longer, something must have gone pretty wrong. 

At least when loading key material with the well-known key types the generalization to potentially handle further types should not produce noticeable overhead. And even for new key types there should be efficient ways of determining which specific (sub-)parser to use.

**Comment by mohd-akram** [comment]
I came across this issue as well while trying to figure out why curl was taking so much time in the handshake. It seems PR #18341 might improve this but it doesn't seem to be released. Will it be released as part of the 3.0 series?

**Comment by t8m** [comment]
> Will it be released as part of the 3.0 series?

Not likely. However we are planning to have 3.1 release relatively soon that will be undergoing FIPS 140-3 validation and it might include performance fixes including this one.

**Comment by l1n** [comment]
Is this going to be resolved by the time that 1.1.1 support is dropped?

**Comment by l1n** [comment]
Checking on this again since we're getting awful close to end of life per the email this morning

**Comment by ThomasChr** [comment]
I'm using Python and this problem has severe performance consequences for multi-threaded Python Code which heavily uses SSL-Requests. We're talking about double the time and up to 15 times the CPU power. See details here, and sorry for the cross post: https://github.com/python/cpython/issues/95031#issuecomment-1748269062

**Comment by colorthoro** [comment]
When will this bug be fixed? It makes Python likes a joke, maybe Python should not trust OpenSSL 3?

**Comment by cyberw** [comment]
Is there any workaround for this issue? 

On windows, setup takes about 4-500ms longer than in previous versions, and as we are doing lots of these it is a major issue..

**Comment by t8m** [comment]
What is the OpenSSL version you're testing with? It should be much improved with 3.2 version.

**Comment by cyberw** [comment]
I'm on an "official" Python 3.12.2 build (installed using pyenv) that comes with OpenSSL 3.0.13. I could probably recompile for my purposes but I can't make all my users to do it, so I'm mostly looking for any kind of workaround without needing a new version.

I guess I could just not support for Python 3.12 on windows, but that would be sad :)

**Comment by gsmethells** [comment]
This is still a major issue years later. The performance degradation is significant. When will resources be applied to fix this issue?

**Comment by t8m** [comment]
Please also note that on Windows (i.e., comment https://github.com/openssl/openssl/issues/18814#issuecomment-2194020984) the biggest part of the performance degradation comes from the use of unbuffered IO when reading PEM files which this PR resolves: https://github.com/openssl/openssl/pull/25716

**Comment by t8m** [comment]
Please note #25716 was merged and it will be included in the 3.5 and 3.4.1 releases.

**Comment by edmorley** [comment]
> Please note [#25716](https://github.com/openssl/openssl/pull/25716) was merged and it will be included in the 3.5 and 3.4.1 releases.

That fix only appears to help with Windows. 

Is work still ongoing for Linux?

It's just we're still seeing significant perf issues with OpenSSL 3 cert loading performance on Linux - particularly for Ubuntu 24.04+, where a new `/usr/lib/ssl/cert.pem` symlink to `/etc/ssl/certs/ca-certificates.crt` has been added due to upstream [Debian bug #805646](https://bugs-devel.debian.org/cgi-bin/bugreport.cgi?bug=805646) - which means [X509_STORE_set_default_paths](https://github.com/openssl/openssl/blob/a1c6e2d1b590dc6a3d2e1c7bd1bf61ffcf854104/crypto/x509/x509_d2.c#L15-L44) is now loading the certs file for the first time (and not just the certs dir `/etc/ssl/certs`). (For more background, see https://github.com/heroku/base-images/pull/344)

**Comment by nhorman** [comment]
@edmorley do you have any data to codify your performance?

I ask because, using the test case outlined in this issue [here](https://github.com/openssl/openssl/issues/18814#issuecomment-2194020984)

on an ubuntu 24.04 vm, I get the following, comparing openssl1.1.1w to the head of the master tree:
```
nhorman@ubuntu:~/test$ ls -l /etc/ssl/certs/ca-certificates.crt 
-rw-r--r-- 1 root root 219342 Feb 16 20:58 /etc/ssl/certs/ca-certificates.crt

nhorman@ubuntu:~/test$ ./measure.py 
Command: loadcert.openssl36 /etc/ssl/certs/ca-certificates.crt Average time over 50 iterations: 0.000212s, 0.000025s CPU time
Command: loadcert.openssl111w /etc/ssl/certs/ca-certificates.crt Average time over 50 iterations: 0.000201s, 0.000023s CPU time
```

Which I think is about 10 usecs worth of difference.  Certainly room for improvement, but not something I would describe as significant, leading me to think you might have something other than the issue described here going on.

**Comment by edmorley** [comment]
Yeah Ruby benchmarks are in https://github.com/heroku/base-images/pull/344

**Comment by nhorman** [comment]
ok, so forgive me here, I've not used the ruby benchmarks before, so I'm learning as I go.  Please review and correct my understanding of the test case you are using:

1) Debian/Ubuntu, in https://bugs-devel.debian.org/cgi-bin/bugreport.cgi?bug=805646 added a symlink to /etc/ssl/certs/ca-certificates.crt at /usr/lib/ssl/cert.pem to force the loading of a trusted root certificate

2) The ruby benchmark as codified by the command:
```
hyperfine -w10 "ruby --disable='gems,did_you_mean,rubyopt' -ropenssl -e 'OpenSSL::X509::Store.new.set_default_paths'" "SSL_CERT_FILE=/dev/null ruby --disable='gems,did_you_mean,rubyopt' -ropenssl -e 'OpenSSL::X509::Store.new.set_default_paths'"
```

Is running the same benchmark twice, once with a 'default' configuration, and again with SSL_CERT_FILE=/dev/null to effectively suppress the loading of the trusted root certificate, the collective output for which is:
```
Benchmark 1: ruby --disable='gems,did_you_mean,rubyopt' -ropenssl -e 'OpenSSL::X509::Store.new.set_default_paths'
  Time (mean ± σ):      33.3 ms ±   2.0 ms    [User: 28.8 ms, System: 4.5 ms]
  Range (min … max):    31.6 ms …  47.6 ms    88 runs

Benchmark 2: SSL_CERT_FILE=/dev/null ruby --disable='gems,did_you_mean,rubyopt' -ropenssl -e 'OpenSSL::X509::Store.new.set_default_paths'
  Time (mean ± σ):      12.3 ms ±   0.6 ms    [User: 8.5 ms, System: 3.9 ms]
  Range (min … max):    11.4 ms …  15.9 ms    220 runs
```


Assuming I have that right, I have two observations:
1) Its not a performance degradation between openssl versions, as this bug describes.  Its probably worth a (potential) separate issue.

2) The benchmark isn't an apples-to-apples comparison.  The debian change, as you noted forces set_default_paths to load an extra file, which appears to (based on strace output) to take about 10k usecs, plus whatever (potentially significant) additional time is needed to parse and decode the certificate. So, unless I'm missing something this isn't a comparison of the same workload.

As for the windows fix, its not really applicable here, as FILE object buffering is on by default in linux, unless you see something that has changed that.

**Comment by edmorley** [comment]
Yeah that's correct.

The issue for us is more that the upstream Debian change now means the "load certs from a file" case is now exercised per-request (depending on the design of end user apps, of which we have no control over as a Platform as a Service), when before only the "load certs from the certs directory" case was. And as such, it's exposing the fact that for OpenSSL 3 the "load certs from a file" case is particularly slow.

This GitHub issue seemed related (to OpenSSL 3 having slow cert loading performance), but perhaps I misread it?

**Comment by edmorley** [comment]
Though other ecosystems are also seeing issues, eg:
https://github.com/python/cpython/issues/95031

**Comment by t8m** [comment]
But this is NOT a performance degradation from 1.1.1 to 3.5/3.6 versions. It is a result of a change in Ubuntu/Debian.

**Comment by nhorman** [comment]
@edmorley I don't think you misread this issue, I only mean to say that this particular issue focused on the fact that cert loading performance was slow on windows because, in openssl 1.1.1 file i/o was buffered, but an intervening change in openssl 3.x switched it to being unbuffered.

The issue you're describing isn't a performance comparison between openssl 1.1.1 and 3.x, but rather a difference in platform configuration that causes more work to be done, for which the fix is going to be something orthogonal to that (either a change to how the platform loads trusted root certificates), or a slow analysis of the certificate loading/decoding path to find performance improvement (which may or may not be possible).  Thats why I suggested a new issue, as whatever those fixes might be is not going to be a change to file i/o buffering settings, as was concluded here.

**Comment by edmorley** [comment]
> But this is NOT a performance degradation from 1.1.1 to 3.5/3.6 versions. It is a result of a change in Ubuntu/Debian.

I think you might have misread what I wrote. I'm aware that the symlink change is a Debian change, what I'm saying is that the symlink change now means the fact that OpenSSL 3 perf is worse than 1.1 is now more of an issue than it was before, hence:

> And as such, it's exposing the fact that for OpenSSL 3 the "load certs from a file" case is particularly slow.

**Comment by nhorman** [comment]
@edmorley Have you shown that Openssl 1.1.1 performance for the load certs from a file case is worse on the latest 3.x than it is on 1.1.1 though?  Your test doesn't show that, it only shows that ruby performance is slower with the debian change than without it.

Mind you, I'm not saying that statement is untrue, only that any performance measurements done in this issue don't address that question, and I'm not sure what the real answer is there.

If I run your benchmark as follows:
```
hyperfine -w20 "LD_LIBRARY_PATH=/home/nhorman/git/worktrees/old ruby --disable='gems,did_you_mean,rubyopt' -ropenssl -e 'OpenSSL::X509::Store.new.set_default_paths'" "LD_LIBRARY_PATH=/home/nhorman/git/openssl/ ruby --disable='gems,did_you_mean,rubyopt' -ropenssl -e 'OpenSSL::X509::Store.new.set_default_paths'"
```

Where my "old" directory houses a copy of openssl-1.1.1w and my "openssl" directory houses a copy of openssl3.6-dev, I get the following results:

```
Benchmark 1: LD_LIBRARY_PATH=/home/nhorman/git/worktrees/old ruby --disable='gems,did_you_mean,rubyopt' -ropenssl -e 'OpenSSL::X509::Store.new.set_default_paths'
  Time (mean ± σ):      34.4 ms ±   2.7 ms    [User: 29.2 ms, System: 5.2 ms]
  Range (min … max):    32.4 ms …  47.6 ms    88 runs

Benchmark 2: LD_LIBRARY_PATH=/home/nhorman/git/openssl/ ruby --disable='gems,did_you_mean,rubyopt' -ropenssl -e 'OpenSSL::X509::Store.new.set_default_paths'
  Time (mean ± σ):      13.9 ms ±   2.1 ms    [User: 8.7 ms, System: 5.2 ms]
  Range (min … max):    12.2 ms …  28.9 ms    224 runs
```

Which suggests to me that openssl on the master branch is currently significantly _faster_ than 1.1.1w was, when comparing a similar workload in which both load the default trust certificate, changing only the underlying openssl library in use

**Comment by edmorley** [comment]
> Have you shown that Openssl 1.1.1 performance for the load certs from a file case is worse on the latest 3.x than it is on 1.1.1 though?

I have not - I had interpreted this issue's title (and the fact it's still open and with many replies) as meaning there was still a known OpenSSL 3 regression. Sorry if that's not the case!

The other difficulty is that we're constrained by the OpenSSL version shipped in the LTS Ubuntu versions - so it may well be that perf is back to (or better than) 1.1.1 performance levels on OpenSSL `master`. For now, we're likely going to just delete the Debian added symlink which triggers the chronic behaviour.

**Comment by nhorman** [comment]
Don't be sorry, it may well be that there is still an opportunity to improve performance here, just not sure what it would be, and is likely not whats covered in this issue (which should probably be closed given that the referenced prs have been merged), but it may be worth opening a separate issue on, as your test case doesn't really cover what was discussed here.

And if you are constrained by what Ubuntu ships in its LTS, yes, thats a separate problem.  It appears that ubuntu 24.04.2 ships with openssl 3.0.13, which definitely had some performance issues that have been addressed in intervening versions.  It may be worth opening a request with ubuntu to update their repository, or offering a newer version via snap.
